ibis/pandas/execution/tests/test_localcontext.py

```diff
@@ -0,0 +1,119 @@
+    assert len(df_all["plain_datetimes_naive"]) == 3
+
+    # with context set, execute produces only rows within context
+    df_within_context = expr.execute(localcontext=context)
+    assert len(df_within_context["plain_datetimes_naive"]) == 1
+
+    del pre_execute.funcs[(PandasTable, PandasClient)]
+    pre_execute.reorder()
+    pre_execute._cache.clear()
+
+
+merge_asof_minversion = pytest.mark.skipif(
+    pd.__version__ < '0.19.2',
+    reason="at least pandas-0.19.2 required for merge_asof",
+)
+
+
+@merge_asof_minversion
+def test_context_adjustment(
+    time_keyed_left, time_keyed_right, time_keyed_df1, time_keyed_df2
+):
+    # pre_execute for trimming table data
+    @pre_execute.register(PandasTable, PandasClient)
+    def execute_table_with_localcontext(
+        op, client, scope=None, localcontext=None, **kwargs
+    ):
+        df = client.dictionary[op.name]
+        try:
+            begin, end = map(pd.to_datetime, localcontext)
+        except ValueError:
+            return {op: df}
+        time_col = "time"
+        return {op: df[df[time_col] >= begin][df[time_col] < end]}
+
+    # define a custom context adjustment rule for asof_join
+    @compute_local_context.register(ops.AsOfJoin, PandasClient)
+    def adjust_context_asof_join(
+        op, client, scope=None, localcontext=None, **kwargs
+    ):
+        new_localcontexts = [
+            localcontext for arg in op.inputs if is_computable_input(arg)
+        ]
+        # right table should look back or forward
+        try:
+            begin, end = map(pd.to_datetime, localcontext)
+            tolerance = op.tolerance
+            if tolerance is not None:
+                timedelta = pd.Timedelta(-tolerance.op().right.op().value)
+                if timedelta <= pd.Timedelta(0):
+                    new_begin = begin + timedelta
+                    new_end = end
+                else:
+                    new_begin = begin
+                    new_end = end + timedelta
+            # right table is the second node in children
+            new_localcontexts[1] = (new_begin, new_end)
+        except ValueError:
+            pass
+        finally:
+            return new_localcontexts
+
+    expr = time_keyed_left.asof_join(
+        time_keyed_right, 'time', by='key', tolerance=2 * ibis.interval(days=1)
+    )[time_keyed_left, time_keyed_right.other_value]
+    context = (3, 4)
+    result = expr.execute(localcontext=context)
+
+    trimmed_df1 = time_keyed_df1[
+        time_keyed_df1["time"] >= pd.to_datetime(context[0])
+    ][time_keyed_df1["time"] < pd.to_datetime(context[1])]
+    trimmed_df2 = time_keyed_df2[
+        time_keyed_df2["time"] >= pd.to_datetime(context[0] - 2)
+    ][time_keyed_df2["time"] < pd.to_datetime(context[1])]
+    expected = pd.merge_asof(
+        trimmed_df1,
+        trimmed_df2,
+        on='time',
+        by='key',
+        tolerance=pd.Timedelta('2D'),
+    )
+    tm.assert_frame_equal(result[expected.columns], expected)
```

Comment:
Why not 

```
tm.assert_frame_equal(result, expected)
```
?

Reply:
Right, they are the same, removed the redundant code.